diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..4c0c6c0
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,20 @@
+set(libpe_Static_SRC
+        error.c
+        hashes.c
+        libfuzzy/edit_dist.c
+        libudis86/decode.c
+        libudis86/syn-att.c
+        libudis86/syn-intel.c
+        misc.c
+        resources.c
+        exports.c
+        imports.c
+        libfuzzy/fuzzy.c
+        libudis86/itab.c
+        libudis86/syn.c
+        libudis86/udis86.c
+        pe.c
+
+        )
+add_library(pe OBJECT ${libpe_Static_SRC})
+target_compile_definitions(pe PUBLIC OBJ)
